scene/main/viewport.cpp

```diff
@@ -1561,44 +1561,15 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 				set_input_as_handled();
 			}
 
-			if (gui.drag_data.get_type() != Variant::NIL && mb->get_button_index() == MouseButton::LEFT) {
+			if (gui.dragging && mb->get_button_index() == MouseButton::LEFT) {
 				// Alternate drop use (when using force_drag(), as proposed by #5342).
-				gui.drag_successful = false;
-				if (gui.mouse_focus) {
-					gui.drag_successful = _gui_drop(gui.mouse_focus, pos, false);
-				}
-
-				gui.drag_data = Variant();
-				gui.dragging = false;
-
-				Control *drag_preview = _gui_get_drag_preview();
-				if (drag_preview) {
-					memdelete(drag_preview);
-					gui.drag_preview_id = ObjectID();
-				}
-				_propagate_viewport_notification(this, NOTIFICATION_DRAG_END);
-				get_base_window()->update_mouse_cursor_shape();
+				_perform_drop(gui.mouse_focus, pos);
 			}
 
 			_gui_cancel_tooltip();
 		} else {
-			if (gui.drag_data.get_type() != Variant::NIL && mb->get_button_index() == MouseButton::LEFT) {
-				gui.drag_successful = false;
-				if (gui.drag_mouse_over) {
-					gui.drag_successful = _gui_drop(gui.drag_mouse_over, gui.drag_mouse_over_pos, false);
-				}
-
-				Control *drag_preview = _gui_get_drag_preview();
-				if (drag_preview) {
-					memdelete(drag_preview);
-					gui.drag_preview_id = ObjectID();
-				}
-
-				gui.drag_data = Variant();
-				gui.dragging = false;
-				gui.drag_mouse_over = nullptr;
```

Comment:
Everything seems to check out from a newbie's perspective - the only thing I noticed is that this line does not appear in the other bit of code replaced by `perform_drop`. With the call to `perform_drop`, both now set `gui.drag_mouse_over` to `nullptr`.

Is there any case where this makes a difference, or was it an oversight from the original implementation?

Reply:
`gui.drag_mouse_over` contains a reference to the `Control`, that the mouse-cursor is currently hovering over and is used during the drop-operation.

It is always set, when the mouse is moved, so drag and drop would also work, in the same way even if it would not be reset in that line. Resetting it to `nullptr` in all cases seemed to be the cleaner way for the implementation.

Reply:
Took some time to look through the mouse event handling code just for fun. What you said makes sense to me. Whenever dragging is happening, `drag_mouse_over` is being set.

I'm guessing everything would work even if `drag_mouse_over` wasn't `nullptr`, since you only check that if you've checked `gui.dragging` anyway. For consistency, I do like that it is `nullptr`, in case someone decides to bypass checking `gui.dragging`.